src/utils/chartJs/CustomTooltip.jsx

```diff
@@ -1,57 +1,85 @@
 import React from 'react';
 import { withStyles } from '@material-ui/core/styles';
 
+const topAligned = {
+  '--trans-y': 'var(--tip-size)',
+  '&::before': { bottom: '100%', borderBottomColor: 'var(--bg-color)' },
+  '&.xright': {
+    '--trans-x': 'calc(-100% + 2 * var(--tip-size))',
+    '&::before': { right: 'var(--tip-size)' },
+  },
+  '&.xcenter': {
+    '--trans-x': '-50%',
+    '&::before': { left: 'calc(50% - var(--tip-size))' },
+  },
+  '&.xleft': {
+    '--trans-x': 'calc(-2 * var(--tip-size))',
+    '&::before': { left: 'var(--tip-size)' },
+  },
+};
 const styles = {
   tooltip: {
+    '--bg-color': 'rgba(0, 0, 0, .8)',
+    '--tip-size': '6px',
     position: 'absolute',
     padding: '6px',
-    background: 'rgba(0, 0, 0, .8)',
+    background: 'var(--bg-color)',
     color: 'white',
     borderRadius: '4px',
     pointerEvents: 'none',
+    transform: 'translate(var(--trans-x), var(--trans-y))',
```

Comment:
**this** is blows my mind!  Let me make sure I get this straight (and only for `x` since `y` is similar)

* the `tooltipModel.xAlign` is used to choose the class (let me call it the "xClass").
* the choice of xClass sets the `--trans-x` variable value
* the `transform` uses `--trans-x` to position the tooltip

Can `tooltipModel.xAlign` set the `transform` x-coordinate directly?  I understand that would mean we still have math in the code, but piping coordinate calculations through class names is convoluted.  

I will say, what you have done is an impressive display of CSS skills. My CSS skills are poor, so you have taught me a few things.  Thank you.  

In general, I appreciate the work you have done in getting the element to position properly, for this reason I am inclined to accept this PR now because it works rather than invest more hours into "simplifying" the code. Even with "simplification" you will still be fighting with the HTML/CSS box model, and the complicated math that it demands, and code that switches based on property names.  Really, not much better.  

There is a better way, [which I have written about](https://github.com/klahnakoski/jsLayout). My code does not work in a JSX React environment (because it manipulates HTML elements directly), but the idea should be translated in to a library that will allow for easier alignment expressions.

For example, Let `p`, `a`, and `t` represent the point, arrow and tooltip respectively:

for top-right positioning 

![image](https://user-images.githubusercontent.com/2334429/55277903-af11b700-52db-11e9-8fdd-3d4b3ce8e3b0.png)

we have 

    p.mc = a.mc
    a.tl = t.bl

more explicitly 

    p.middle = a.middle
    p.center = a.center
    a.top = t.bottom
    a.left = t.left

this ignores the `cornerspace`, but that can be solved with more `divs`, or using `box-sizing` to clarify the alignment.  

That is all a dream for the future, but I hope you keep it in mind when you are positioning elements.

Reply:
> * the `tooltipModel.xAlign` is used to choose the class (let me call it the "xClass").
> * the choice of xClass sets the `--trans-x` variable value
> * the `transform` uses `--trans-x` to position the tooltip

Yup, that's exactly what it does, as well as to adjust the triangle tip (`::before`).

> Can `tooltipModel.xAlign` set the `transform` x-coordinate directly? I understand that would mean we still have math in the code, but piping coordinate calculations through class names is convoluted.

You do love perfecting stuff :sweat_smile: . Yup yup, as I mentioned in a previous comment [JSS Dynamic values](https://cssinjs.org/react-jss?v=v10.0.0-alpha.16#dynamic-values) would help us move all the style related stuff outside the component, but I wasn't able to make it work in my first try, I'll keep working on it in my spare time, don't worry.

>  My CSS skills are poor, so you have taught me a few things. Thank you.

Thank YOU! As I have mentioned many times before, all your previous reviews, your comments, and the way you look for perfection is what makes me learn more and search for new ways to improve what's been done, and I freaking enjoy that.

Checking your method...

Reply:
Wow just wow. I just checked your jsLayout library, and I'm surprised that it doesn't have the popularity that should deserve. It gives you a 'ping' and 'elastic' feeling. Like a canvas api with steroids, for DOM elements. Love it.


> That is all a dream for the future, but I hope you keep it in mind when you are positioning elements.

Yup yup, It's already stuck in my mind and I'll see where and how your idea can be applied in this project or others. Thank you!!